File: expo/app/src/flow/use-smile-id-sample-flow-preflight.ts
import {
  smileIDSampleRequirementFrom,
  type UseSmileIDSampleUserDetailsRequirement,
} from '@smileid/sample-ui';
import {
  UseSmileIDFlowBuilder,
  type ValidationState,
  type UseSmileIDValidationException,
} from '@smileid/usesmileid';

import { smileIDSampleApplying } from './use-smile-id-sample-flow-builder-config';
import {
  smileIDSampleSnapshotSession,
  type UseSmileIDSampleFlowLaunchSnapshot,
} from './use-smile-id-sample-flow-launch-snapshot';

/// What the gate decided, and so where the journey goes instead of the SDK.
export type UseSmileIDSampleFlowPreflight =
  | { readonly kind: 'ready' }
  /// The forms can resolve it.
  | { readonly kind: 'needsDetails'; readonly issues: readonly UseSmileIDValidationException[] }
  /// Only a new token resolves it, so the journey goes back to the scanner rather than to a form.
  | { readonly kind: 'needsSession' }
  /// No form can resolve it, and it must still never reach the SDK.
  | { readonly kind: 'misconfigured'; readonly issues: readonly UseSmileIDValidationException[] };

/// The entry gate: the SDK's non-throwing pre-flight plus its per-payload validators.
export const smileIDSamplePreflight = (
  snapshot: UseSmileIDSampleFlowLaunchSnapshot,
): UseSmileIDSampleFlowPreflight => {
  // Ahead of the payloads, because no form fixes a session that has run out.
  if (snapshot.sessionExpired) return { kind: 'needsSession' };
  const builder = new UseSmileIDFlowBuilder();
  smileIDSampleApplying(builder, snapshot);
  const requirement = smileIDSampleRequirementFrom(smileIDSampleSnapshotSession(snapshot)?.bindings);

  // Payloads before the builder's verdict: a form can fix what was typed, not how this built it.
  const checks: ValidationState[] = [];
  if (builder.userDetails !== undefined) {
    // The public validator takes no token, so the bindings come off what it reports instead.
    checks.push(minus(builder.validateUserDetails(builder.userDetails), requirement));
  }
  if (builder.biometricKYCParams !== undefined)
    checks.push(builder.validateBiometricKYCParams(builder.biometricKYCParams));
  if (builder.enhancedKYCParams !== undefined)
    checks.push(builder.validateEnhancedKYCParams(builder.enhancedKYCParams));
  if (builder.documentVerificationParams !== undefined)
    checks.push(builder.validateDocumentVerificationParams(builder.documentVerificationParams));
  if (builder.enhancedDocumentVerificationParams !== undefined)
    checks.push(
      builder.validateEnhancedDocumentVerificationParams(builder.enhancedDocumentVerificationParams),
    );

  const payloadIssues = checks.flatMap((state) => (state.valid ? [] : state.issues));
  if (payloadIssues.length > 0) return { kind: 'needsDetails', issues: payloadIssues };

  const verdict = builder.validate();
  return verdict.valid ? { kind: 'ready' } : { kind: 'misconfigured', issues: verdict.issues };
};

/// Drops the issues the token already answers; every other rule the SDK applies still stands.
const minus = (state: ValidationState, requirement: UseSmileIDSampleUserDetailsRequirement): ValidationState => {
  if (state.valid) return state;
  const outstanding = state.issues.filter((issue) => !covers(requirement, issue));
  return outstanding.length === 0 ? { valid: true } : { ...state, issues: outstanding };
};

const covers = (requirement: UseSmileIDSampleUserDetailsRequirement, issue: UseSmileIDValidationException): boolean => {
  const details = issue.errorDetails as { fieldName?: unknown; reason?: unknown } | undefined;
  switch (details?.fieldName) {
    case 'userDetails.givenNames':
      return !requirement.firstName;
    case 'userDetails.lastName':
      return !requirement.lastName;
    // The contact rule is reported against the object, so its reason is what identifies it.
    case 'userDetails':
      return !requirement.contact && String(details.reason).toLowerCase().includes('email');
    default:
      return false;
  }
};

File: flutter/app/lib/src/flow/use_smileid_sample_flow_preflight.dart
import 'package:usesmileid/usesmileid.dart';

import 'use_smileid_sample_flow_builder_config.dart';
import 'use_smileid_sample_flow_launch_snapshot.dart';

/// What the gate decided, and so where the journey goes instead of the SDK.
sealed class UseSmileIDSampleFlowPreflight {
  const UseSmileIDSampleFlowPreflight();
}

/// Nothing stands between this snapshot and the SDK.
class UseSmileIDSampleFlowReady extends UseSmileIDSampleFlowPreflight {
  /// No fields: a ready gate has nothing to report.
  const UseSmileIDSampleFlowReady();
}

/// The forms can resolve it.
class UseSmileIDSampleFlowNeedsDetails extends UseSmileIDSampleFlowPreflight {
  /// [issues] is what the SDK's own validators reported.
  const UseSmileIDSampleFlowNeedsDetails(this.issues);

  /// What the payload validators rejected.
  final List<UseSmileIDValidationException> issues;
}

/// No form can resolve it, and it must still never reach the SDK.
class UseSmileIDSampleFlowMisconfigured extends UseSmileIDSampleFlowPreflight {
  /// [issues] is what `validate()` reported about the builder itself.
  const UseSmileIDSampleFlowMisconfigured(this.issues);

  /// What the builder validator rejected.
  final List<UseSmileIDValidationException> issues;
}

/// The entry gate: the SDK's non-throwing pre-flight plus its per-payload validators.
UseSmileIDSampleFlowPreflight useSmileIDSamplePreflight(
  UseSmileIDSampleFlowLaunchSnapshot snapshot,
) {
  final UseSmileIDFlowBuilder builder = UseSmileIDFlowBuilder();
  useSmileIDSampleApplying(builder, snapshot);

  // Payloads before the builder's verdict: a form can fix what was typed, not how this built it.
  final List<ValidationState> payloadChecks = <ValidationState>[
    if (builder.userDetails case final UserDetails details)
      builder.validateUserDetails(details),
    if (builder.biometricKYCParams case final BiometricKYCParams params)
      builder.validateBiometricKYCParams(params),
    if (builder.enhancedKYCParams case final EnhancedKYCParams params)
      builder.validateEnhancedKYCParams(params),
    if (builder.documentVerificationParams
        case final DocumentVerificationParams params)
      builder.validateDocumentVerificationParams(params),
    if (builder.enhancedDocumentVerificationParams
        case final EnhancedDocumentVerificationParams params)
      builder.validateEnhancedDocumentVerificationParams(params),
  ];
  final List<UseSmileIDValidationException> payloadIssues =
      <UseSmileIDValidationException>[
        for (final ValidationState state in payloadChecks)
          if (state case final ValidationStateInvalid invalid)
            ...invalid.issues,
      ];
  if (payloadIssues.isNotEmpty) {
    return UseSmileIDSampleFlowNeedsDetails(payloadIssues);
  }
  return switch (builder.validate()) {
    ValidationStateValid() => const UseSmileIDSampleFlowReady(),
    final ValidationStateInvalid invalid => UseSmileIDSampleFlowMisconfigured(
      invalid.issues,
    ),
  };
}
